sparse/_common.py

```diff
@@ -0,0 +1,379 @@
+    format : str, optional
+        A format string.
+    compressed_axes : iterable, optional
+        The axes to compress if returning a GCXS array.
+
+    Returns
+    -------
+    I : [COO, GCXS, DOK] array of shape (N, M)
+        An array where all elements are equal to zero, except for the `k`-th
+        diagonal, whose values are equal to one.
+
+    Examples
+    --------
+    >>> eye(2, dtype=int).todense()  # doctest: +NORMALIZE_WHITESPACE
+    array([[1, 0],
+           [0, 1]])
+    >>> eye(3, k=1).todense()  # doctest: +SKIP
+    array([[0., 1., 0.],
+           [0., 0., 1.],
+           [0., 0., 0.]])
+    """
+    from sparse import COO
+
+    if M is None:
+        M = N
+
+    N = int(N)
+    M = int(M)
+    k = int(k)
+
+    data_length = min(N, M)
+
+    if k > 0:
+        data_length = max(min(data_length, M - k), 0)
+        n_coords = np.arange(data_length, dtype=np.intp)
+        m_coords = n_coords + k
+    elif k < 0:
+        data_length = max(min(data_length, N + k), 0)
+        m_coords = np.arange(data_length, dtype=np.intp)
+        n_coords = m_coords - k
+    else:
+        n_coords = m_coords = np.arange(data_length, dtype=np.intp)
+
+    coords = np.stack([n_coords, m_coords])
+    data = np.array(1, dtype=dtype)
+
+    return COO(
+        coords, data=data, shape=(N, M), has_duplicates=False, sorted=True
+    ).asformat(format, compressed_axes=compressed_axes)
+
+
+def full(shape, fill_value, dtype=None, format="coo", compressed_axes=None):
+    """Return a COO array of given shape and type, filled with `fill_value`.
+
+    Parameters
+    ----------
+    shape : int or tuple of ints
+        Shape of the new array, e.g., ``(2, 3)`` or ``2``.
+    fill_value : scalar
+        Fill value.
+    dtype : data-type, optional
+        The desired data-type for the array. The default, `None`, means
+        `np.array(fill_value).dtype`.
+    format : str, optional
+        A format string.
+    compressed_axes : iterable, optional
+        The axes to compress if returning a GCXS array.
+    Returns
+    -------
+    out : [COO, GCXS, DOK]
+        Array of `fill_value` with the given shape and dtype.
+
+    Examples
+    --------
+    >>> full(5, 9).todense()  # doctest: +NORMALIZE_WHITESPACE
+    array([9, 9, 9, 9, 9])
+
+    >>> full((2, 2), 9, dtype=float).todense()  # doctest: +SKIP
+    array([[9., 9.],
+           [9., 9.]])
+    """
+    from sparse import COO
+
+    if dtype is None:
+        dtype = np.array(fill_value).dtype
+    if not isinstance(shape, tuple):
+        shape = (shape,)
+    data = np.empty(0, dtype=dtype)
+    coords = np.empty((len(shape), 0), dtype=np.intp)
+    return COO(
+        coords,
+        data=data,
+        shape=shape,
+        fill_value=fill_value,
+        has_duplicates=False,
+        sorted=True,
+    ).asformat(format, compressed_axes=compressed_axes)
+
+
+def full_like(a, fill_value, dtype=None, format="coo", compressed_axes=None):
```

Comment:
Can we make the default here be `format=None`? These changes are so that format can be inferred from `a` as well.